src/operations/update.ts

```diff
@@ -41,107 +46,260 @@ export interface UpdateResult {
-        upsertedCount: Array.isArray(r.upserted) && r.upserted.length ? r.upserted.length : 0,
-        matchedCount: Array.isArray(r.upserted) && r.upserted.length > 0 ? 0 : r.n
+        modifiedCount: res.nModified != null ? res.nModified : res.n,
+        upsertedId:
+          Array.isArray(res.upserted) && res.upserted.length > 0 ? res.upserted[0]._id : null,
+        upsertedCount: Array.isArray(res.upserted) && res.upserted.length ? res.upserted.length : 0,
+        matchedCount: Array.isArray(res.upserted) && res.upserted.length > 0 ? 0 : res.n
       });
     });
   }
 }
 
 /** @internal */
-export class UpdateManyOperation extends CommandOperation<UpdateResult> {
-  options: UpdateOptions;
-  collection: Collection;
-  filter: Document;
-  update: Document;
-
+export class UpdateManyOperation extends UpdateOperation {
   constructor(collection: Collection, filter: Document, update: Document, options: UpdateOptions) {
-    super(collection, options);
+    super(
+      collection.s.namespace,
+      [makeUpdateOperation(filter, update, { ...options, multi: true })],
+      options
+    );
 
-    this.options = options;
-    this.collection = collection;
-    this.filter = filter;
-    this.update = update;
+    if (!hasAtomicOperators(update)) {
+      throw new TypeError('Update document requires atomic operators');
+    }
+  }
+
+  execute(
+    server: Server,
+    session: ClientSession,
+    callback: Callback<UpdateResult | Document>
+  ): void {
+    super.execute(server, session, (err, res) => {
+      if (err || !res) return callback(err);
+      if (typeof this.explain !== 'undefined') return callback(undefined, res);
+
+      // TODO: copied from `updateDocuments`, not sure if these are needed anymore
+      if (res.code) return callback(new MongoError(res));
+      if (res.writeErrors) return callback(new MongoError(res.writeErrors[0]));
+
+      callback(undefined, {
+        acknowledged: this.writeConcern?.w !== 0 ?? true,
+        modifiedCount: res.nModified != null ? res.nModified : res.n,
+        upsertedId:
+          Array.isArray(res.upserted) && res.upserted.length > 0 ? res.upserted[0]._id : null,
+        upsertedCount: Array.isArray(res.upserted) && res.upserted.length ? res.upserted.length : 0,
+        matchedCount: Array.isArray(res.upserted) && res.upserted.length > 0 ? 0 : res.n
+      });
+    });
+  }
+}
+
+/** @public */
+export interface ReplaceOptions extends CommandOperationOptions {
+  /** If true, allows the write to opt-out of document level validation */
+  bypassDocumentValidation?: boolean;
+  /** Specifies a collation */
+  collation?: CollationOptions;
+  /** Specify that the update query should only consider plans using the hinted index */
+  hint?: string | Document;
+  /** When true, creates a new document if no document matches the query */
+  upsert?: boolean;
+
+  // FIXME:
+  multi?: boolean;
```

Comment:
I think the fix here would be to remove multi from the interface and instead have like `ReplaceOptions & {multi: boolean}` since it's like an internal thing, does that sound right?
We don't have to do that here but lets leave that in the fixme comment for the next dev

Reply:
oops, this was just cruft. I actually did the `& { multi: boolean }` trick already in `makeUpdateOperation` below!